pkg/dockerregistry/server/prune/restore.go

```diff
@@ -0,0 +1,380 @@
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.cache == nil {
+		s.cache = make(map[digest.Digest]bool)
+	}
+
+	if v, ok := s.cache[dgst]; ok {
+		return v, nil
+	}
+
+	_, err := s.statter.Stat(ctx, dgst)
+
+	if err != nil && err != distribution.ErrBlobUnknown {
+		return false, err
+	}
+
+	s.cache[dgst] = err == nil
+
+	return s.cache[dgst], nil
+}
+
+// ErrImage defines the error associated with the digest
+type ErrImage struct {
+	Digest  digest.Digest
+	Problem error
+}
+
+// Error implements the error interface
+func (err ErrImage) Error() string {
+	return err.Problem.Error()
+}
+
+// Fsck validates or recovers database based on storage
+type Fsck struct {
+	Ctx        context.Context
+	Client     client.Interface
+	Registry   distribution.Namespace
+	ServerAddr string
+	Restore    Restore
+}
+
+func (r *Fsck) checkImage(image *imageapiv1.Image, blobStatter *statter) error {
+	if !imagestream.IsImageManaged(image) {
+		return nil
+	}
+
+	if err := originutil.ImageWithMetadata(image); err != nil {
+		return fmt.Errorf("error getting image metadata: %s", err)
+	}
+
+	imageDigest, err := digest.ParseDigest(image.Name)
+	if err != nil {
+		return fmt.Errorf("bad image name %q: %s", image.Name, err)
+	}
+
+	exists, err := blobStatter.exists(r.Ctx, imageDigest)
+	if err != nil {
+		return err
+	} else if !exists {
+		return &ErrImage{
+			Digest:  imageDigest,
+			Problem: distribution.ErrBlobUnknown,
+		}
+	}
+
+	if image.DockerImageManifestMediaType == schema2.MediaTypeManifest {
+		meta, ok := image.DockerImageMetadata.Object.(*dockerapiv10.DockerImage)
+		if ok {
+			configDigest, err := digest.ParseDigest(meta.ID)
+			if err != nil {
+				return fmt.Errorf("image %q: bad config %q: %s", imageDigest, meta.ID, err)
+			}
+
+			exists, err := blobStatter.exists(r.Ctx, configDigest)
+			if err != nil {
+				return fmt.Errorf("image %q: config %q: %s", imageDigest, configDigest, err)
+			} else if !exists {
+				return &ErrImage{
+					Digest:  configDigest,
+					Problem: distribution.ErrBlobUnknown,
+				}
+			}
+		}
+	}
+
+	for _, layer := range image.DockerImageLayers {
+		layerDigest, err := digest.ParseDigest(layer.Name)
+		if err != nil {
+			return fmt.Errorf("image %q: bad layer %q: %s", imageDigest, layer.Name, err)
+		}
+
+		exists, err := blobStatter.exists(r.Ctx, layerDigest)
+		if err != nil {
+			return fmt.Errorf("image %q: layer %q: %s", imageDigest, layer.Name, err)
+		} else if !exists {
+			return &ErrImage{
+				Digest:  layerDigest,
+				Problem: distribution.ErrBlobUnknown,
+			}
+		}
+	}
+
+	return nil
+}
+
+// Database checks metadata in the database
+func (r *Fsck) Database(namespace string) error {
+	listIS, err := r.Client.ImageStreams(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		if namespace == metav1.NamespaceAll {
+			namespace = "all"
+		}
+		return fmt.Errorf("failed to list image streams in the %s namespace(s): %s", namespace, err)
```

Comment:
why (s) ?  seems like %s is only ever going to be a single namespace name.

Reply:
@bparees Because `failed to list image streams in the foo namespace` or `failed to list image streams in the all namespaces`. No ?

Reply:
Hm, well it's a bit awkward("in the all namespaces"), how about rewording to:
`failed to list image streams in namespace(s): %s, error: %s`

Reply:
@legionus you may have missed this comment

Reply:
@bparees oops. sorry.